<h1> Soal 1: Pemahaman Tentang Model Evaluasi</h1>

Jawab pertanyaan di bawah ini dengan bahasa masing-masing?

1. Apa perbedaan antara data latih, data validasi, dan data test?
2. Bagaimana cara kita menilai performa suatu model?
3. Apa itu Confusion Matrix? Jelaskan secara lengkap!
4. Apa itu Classification Report dari sklearn?

Jawab:
1. Apa perbedaan antara data latih, data validasi, dan data test?
- data latih adalah data yang digunakan untuk melatih model machine learning
- data validasi adalah data yang digunakan untuk memvalidasi (evaluasi) model machine learning, mengatur parameter atau mencari parameter yang optimal, dan mencegah overfitting
- data test adalah data yang digunakan untuk mengetest model machine learning yang sudah selesai dilatih (mengestimasi error dari model / tingkat keakuratannya)
------
2. Menilai performa model dapat dilakukan dengan fungsi metrics, macam-macam metrics diantaranya
- Accuracy score -> jumlah prediksi yang benar / jumlah data
- Confusion matrix
- AUC
- Log-loss
-------
3. Confusion matrix adalah salah satu metrics untuk mengevaluasi performa model machine learning. Umumnya digunakan jika terdapat class imbalance. Dalam confusion matrix ada 4 kategori tebakan
- tebakan benar, nilainya positif (TP)
- tebakan salah, nilainya positif (FP)
- tebakan benar, nilainya negatif (TN)
- tebakan salah, nilainya negatif (FN)

dan ada 2 tipe error
- error ketika menebak positif -> Type 1 error
- error ketika menebak negatif -> Tyoe 2 error
--------
4. Classification report adalah salah satu fungsi dari sklearn yang dapat digunakan untuk mereport hasil confusion matrix dalam bentuk precision, recall, dan f1-score

---

<h1>Soal 2: Aplikasi Model Evaluasi</h1>

Kali ini kita akan menggunakan data untuk memprediksi kelangsungan hidup pasien yang telah mengalami operasi payudara. Dengan informasi yang dimiliki terkait pasien, kita akan membuat model untuk memprediksi apakah pasien akan bertahan hidup dalam waktu lebih dari 5 tahun atau tidak.

Lebih Lengkapnya kalian bisa membaca informasi tentang dataset di link berikut: https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.names

Buat model Klasifikasi (Model/Algoritma Bebas) untuk memprediksi status pasien dengan ketentuan sebagai berikut:

1. Bagi kedua data ini menjadi data training dan data test dengan test_size=0.25.
3. Pelajar tentang metrics roc_auc_score kemudian buatlah model dan evaluasi dengan menggunakan teknik cross-validation dengan scoring 'roc_auc'. Baca https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html untuk menggunakan metric roc_auc saat cross-validation.
3. Berapa score rata2 dari model dengan teknik cross-validation tersebut?
4. Prediksi data test dengan model yang telah kalian buat!
5. Bagaimana hasil confusion matrix dari hasil prediksi tersebut?
6. Bagaimana classification report dari hasil prediksi tersebut?
5. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
6. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?

In [3]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [4]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [5]:
df['survival_status'].value_counts()

1    225
2     81
Name: survival_status, dtype: int64

In [11]:
# Code here
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, classification_report

# split data features and target
x = df.drop('survival_status', axis=1)
y = df.survival_status

# split data train and data test
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.25)

# create an object model and choose hyperparameter value
knn = KNeighborsClassifier(n_neighbors=6)

# fit / learning process
knn.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=6)

In [12]:
# cross validation
score = cross_val_score(knn, x, y, cv=10, scoring='roc_auc')
print('score rata-rata dari cross validation: ',score.mean())

score rata-rata dari cross validation:  0.6705561045234958


In [13]:
# predict
y_pred = knn.predict(x_test)

In [14]:
# confusion matrix
conf = confusion_matrix(y_test, y_pred, labels=[1, 2])
conf

array([[53,  5],
       [17,  2]], dtype=int64)

In [15]:
# classification report
class_report = classification_report(y_test, y_pred)
print(class_report)

              precision    recall  f1-score   support

           1       0.76      0.91      0.83        58
           2       0.29      0.11      0.15        19

    accuracy                           0.71        77
   macro avg       0.52      0.51      0.49        77
weighted avg       0.64      0.71      0.66        77



- status positif -> mampu bertahan lebih dari 5 tahun
- status negatif -> tidak mampu bertahan hidup lebih dari 5 tahun

---
Dari nilai presisi (1), kemampuan model dalam memprediksi pasien berstatus positif adalah 0.76.
Dari nilai presisi (2), kemampuan model dalam memprediksi pasien berstatus negatif adalah 0.29. Artinya model tidak mampu mendeteksi pasien negatif dengan baik, dimana banyak pasien berstatus negatif malah terdeteksi positif.

true positive rate = x, x=1 setiap orang yang positif diprediksi dengan benar
false positive rate= x, x=1 setiap orang yang negatif diprediksi positif (alias salah kabeh)

If correctly identifying positives is important for us, then we should choose a model with higher Sensitivity. However, if correctly identifying negatives is more important, then we should choose specificity as the measurement metric.

---

<h1> Soal 3: Pemahaman Tentang Model Selection</h1>

Jelaskan dengan bahasa sendiri!

1. Apa itu Bias dan Variance?
2. Apa itu Overfitting dan Underfitting?
3. Apa yang bisa kita lakukan untuk mengatur kompleksitas dari model?
4. Bagaimana model yang baik?
5. Kapan kita menggunakan GridSearchcv dan kapan menggunakan RandomizedSearchCV?

Jawab:
1. Bias adalah perbedaan hasil prediksi dengan nilai yang sebenarnya, terjadi karena model yang terlalu simpel. Variansi adalah nilai tentang variasi/persebaran data atau variasi prediksi model terhadap data yang diberikan, terjadi karena model terlalu kompleks


2. Overfitting adalah kondisi ketika hasil prediksi sangat mendekati nilai aktual, sehingga model tidak general dan tidak bisa digunakan untuk data lain. Hal ini terjadi karena model terlalu kompleks dan terlalu memperhatikan/mempelajari data training, mengambil detail dan juga noise data training. Underfitting adalah kondisi ketika hasil prediksi sangat jauh dari nilai aktual. Hal ini terjadi karena model terlalu simpel.


3. Mengatur kompleksitas model dengan mengatur hyperparameter. Jika ingin menurunkan kompleksitas, maka turunkan jumlah parameter yang dipertimbangkan dalam model, berlaku sebaliknya


4. Model yang baik adalah model yang generaliasi yang bagus, memiliki bias yang kecil dan variance yang kecil, sehingga model tidak akan underfitting atau overfitting.


5. GridSearchCV dapat digunakan ketika hyperparameter yang ditentukan tidak terlalu banyak dan dataset tidak terlalu besar, sedangkan RandomizedSearchCV digunakan ketika terdapat banyak hyperparameter yang ditentukan serta ukuran dataset besar.

---

<h1> Soal 4: Aplikasi Model Selection</h1>

1. Bagi kedua data berikut ini menjadi data training dan data test dengan test_size=0.25.
2. Gunakan algoritma KNN sebagai model classifier.
3. Gunakan fungsi GridSearchCV untuk hyperparameter tuning dan model selection.
4. jumlah fold bebas!, gunakan scoring 'roc_auc'
5. Definisikan kombinasi hyperparameter untuk model selection dengan GridSearchCV. kombinasi Hyperparameter bebas, baca lagi dokumentasi KNN di link berikut https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html untuk memahami lagi jenis2 hyperparameter di algorithma KNN.
6. Latih model terhadap data training.
7. Apa hyperparameter terbaik untuk kombinasi hyperparameter kalian?
8. Berapa score validasi terbaik dari model tersebut?
9. Prediksi probabilitasi output dari model yang telah di buat terhadap data test. note : gunakan method .predict_proba() untuk menghasilkan output probabilitas
10. Perhatikan bahwa hasil prediksi ada 2, dimana masing2 adalah nilai probabilitas untuk setiap class label. Ambil nilai probabilitas pasien phositive meninggal dalam waktu kurang dari 5 tahun. note : gunakan bantuan attirubte .classes_ untuk mengetahui urutan label dari hasil prediksi probabilitas.
11. Berapa nilai score roc_auc untuk data test?
12. Apakah model anda termasuk baik, overtting, atau underfitting?
13. Ulangi tahap di atas namun kali ini menggunakan algoritma DecisionTreeClassifier dan kalian bisa menggunakan RandomizedSearchCV apabila process training lama. pelajari algoritma DecisionTreeClassifier di linkberikut: https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html?highlight=decisiontreeclassifier#sklearn.tree.DecisionTreeClassifier
14. Bandingkan scorenya dengan Algoritma KNN, mana yang lebih baik?

Note : Data Science adalah experiment, sangat di dimungkinkan memerlukan beberapa kali percobaan untuk mendapatkan hasil yang terbaik! Happy Coding :)

In [16]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [17]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [20]:
# Code here
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score, cross_validate
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

# split data features and target
x2 = df.drop('survival_status', axis=1)
y2 = df.survival_status

# 1. split data train and data test
x_train, x_test, y_train, y_test = train_test_split(x2,y2, test_size=0.25)

In [21]:
# 2. pilih model
knn2 = KNeighborsClassifier()

# 3 & 4 & 5. hypertuning & model selection
param_grid = {'n_neighbors': np.arange(5,50), 'weights':['uniform', 'distance']}
knn_gscv = GridSearchCV(knn2, param_grid=param_grid, scoring='roc_auc', cv=10)

In [22]:
# 6. latih model / fitting
knn_gscv.fit(x_train, y_train)

GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
       22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
       39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]),
                         'weights': ['uniform', 'distance']},
             scoring='roc_auc')

In [23]:
# 7. parameter terbaik
knn_gscv.best_params_

{'n_neighbors': 43, 'weights': 'distance'}

In [24]:
# 8. score terbaik
knn_gscv.best_score_

0.7470238095238095

In [25]:
# 9. probabilitas output dari model
# predict() -> cuma output prediksi binari 1 atau 0
# predict_proba() -> probabilitas dari prediksi, seberapa besar probability dari prediksi 1 atau 0

y_pred_proba = knn_gscv.predict_proba(x_test)
y_pred_proba[:5]

array([[0.83258183, 0.16741817],
       [1.        , 0.        ],
       [0.39332346, 0.60667654],
       [0.74524505, 0.25475495],
       [0.83840279, 0.16159721]])

In [26]:
# 10 Ambil nilai probabilitas pasien phositive meninggal dalam waktu kurang dari 5 tahun
# survival = 1 -> the patient survived 5 years or longer
# survival = 2 -> the patient died within 5 years

survived_probs = y_pred_proba[:,0]
died_probs = y_pred_proba[:,1]

print('Probabilitas pasien positif meninggal dalam waktu kurang dari 5 tahun: \n', died_probs)

Probabilitas pasien positif meninggal dalam waktu kurang dari 5 tahun: 
 [0.16741817 0.         0.60667654 0.25475495 0.16159721 0.
 0.12443936 0.         0.         0.10119795 1.         0.11937843
 0.16602641 0.14435883 0.25226128 0.2701761  0.61744757 0.26118937
 0.09580763 0.0403196  0.12501726 0.15150983 0.22470571 0.17031635
 0.0962158  0.2918312  0.28986706 0.64222666 0.50353575 0.52120134
 0.14852859 0.2297855  0.27684358 1.         0.16365024 0.25322901
 0.23951697 0.25921231 0.27836434 0.60465297 1.         0.
 0.1681527  0.19743848 0.2473377  0.22428146 0.21183707 0.19140368
 0.53104928 0.51919309 0.23401829 0.32558806 0.19393261 0.49724937
 0.26975634 0.19701459 0.12162815 0.17652051 0.04026605 0.
 0.1488632  0.48828038 1.         0.67655112 0.27578673 0.49809496
 0.         0.20990992 0.20796145 0.17816932 0.06297444 0.26606695
 0.24760917 0.15607544 0.0883287  0.         0.51992519]


In [28]:
# 11  nilai score roc_auc untuk data test
survived_test_auc_score = roc_auc_score(y_test, survived_probs)
died_test_auc_score = roc_auc_score(y_test, died_probs)

print('Survived more than 5 years: ROC AUC = %.3f' % (survived_test_auc_score))
print('Died within 5 years: ROC AUC = %.3f' % (died_test_auc_score))

Survived more than 5 years: ROC AUC = 0.386
Died within 5 years: ROC AUC = 0.614


12. Apakah model anda termasuk baik, overtting, atau underfitting?

Dari metrics didapat score roc_auc sebesar 0.614. Nilai AOC = 0.614 mendekati 0.5 yang menandakan bahwa model tidak terlalu bisa membedakan antara pasien yang tidak dapat bertahan (died) dan yang dapat bertahan hidup (survived). Model tidak overfitting, tetapi cenderung underfitting. 

### Decision tree classifier

13. Gunakan algoritma DecisionTreeClassifier dan kalian bisa menggunakan RandomizedSearchCV

In [29]:
# pilih model
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier()

# hyperparameter tuning
param_grid = {'criterion':['gini','entropy'], 'max_depth':np.arange(1,10), 'min_samples_leaf':np.arange(2,30), 'random_state':np.arange(2,30)}
dtc_gscv = RandomizedSearchCV(dtc, param_grid, scoring='roc_auc', cv=10)

# learning
dtc_gscv.fit(x_train, y_train)

RandomizedSearchCV(cv=10, estimator=DecisionTreeClassifier(),
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': array([1, 2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_leaf': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]),
                                        'random_state': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29])},
                   scoring='roc_auc')

In [30]:
# best parameter & score
print('best DecisionTreeClassifier model parameter: \n', dtc_gscv.best_params_,'\n')
print('best DecisionTreeClassifier model score: \n', dtc_gscv.best_score_)

best DecisionTreeClassifier model parameter: 
 {'random_state': 8, 'min_samples_leaf': 15, 'max_depth': 4, 'criterion': 'gini'} 

best DecisionTreeClassifier model score: 
 0.7426207983193278


In [31]:
# predict probability
y_pred2_proba = dtc_gscv.predict_proba(x_test)

In [32]:
survived_probs2 = y_pred2_proba[:,0]
died_probs2 = y_pred2_proba[:,1]

print('Probabilitas pasien positif meninggal dalam waktu kurang dari 5 tahun:\n {}'.format(died_probs2))

Probabilitas pasien positif meninggal dalam waktu kurang dari 5 tahun:
 [0.04545455 0.31034483 0.77272727 0.45454545 0.         0.
 0.04545455 0.         0.26666667 0.17647059 0.3125     0.31034483
 0.31034483 0.04545455 0.45454545 0.3125     0.77272727 0.45454545
 0.08571429 0.08571429 0.         0.         0.3125     0.3125
 0.17647059 0.17647059 0.17647059 0.77272727 0.45454545 0.77272727
 0.         0.58823529 0.58823529 0.3125     0.04545455 0.26666667
 0.45454545 0.45454545 0.45454545 0.77272727 0.31034483 0.3125
 0.45454545 0.04545455 0.45454545 0.31034483 0.26666667 0.31034483
 0.77272727 0.77272727 0.         0.45454545 0.26666667 0.17647059
 0.04545455 0.31034483 0.         0.31034483 0.08571429 0.04545455
 0.         0.45454545 0.77272727 0.77272727 0.3125     0.17647059
 0.         0.3125     0.         0.26666667 0.08571429 0.3125
 0.45454545 0.04545455 0.17647059 0.         0.77272727]


In [33]:
# test score - using ROC AUC
survived_auc2 = roc_auc_score(y_test, survived_probs2)
died_auc2 = roc_auc_score(y_test, died_probs2)

print('Survived more than 5 years: ROC AUC = %.3f' % (survived_auc2))
print('Died within 5 years: ROC AUC = %.3f' % (died_auc2))

Survived more than 5 years: ROC AUC = 0.402
Died within 5 years: ROC AUC = 0.598


- Specifity = true negative rate
- 1 - specifity = false positive rate
- AUC ROC indicates how well the probabilities from the positive classes are separated from the negative classes.

14. Bandingkan scorenya dengan Algoritma KNN

Kedua model tidak mampu membedakan label dengan cukup baik. Performa KNN dalam mendeteksi pasien yang tidak dapat bertahan hidup (AUC=0.614) hampir sama dengan Decision Tree (AUC=0.598).

Begitupun dalam mendeteksi pasien yang mampu bertahan hidup, performa KNN (AUC=0.386) tidak jauh berbeda dengan performa Decision Tree (AUC=0.402)